zebra-state/src/service/non_finalized_state.rs

```diff
@@ -199,12 +202,20 @@ impl NonFinalizedState {
             &parent_chain.spent_utxos,
             finalized_state,
         )?;
+
         check::prepared_block_commitment_is_valid_for_chain_history(
             &prepared,
             self.network,
             &parent_chain.history_tree,
         )?;
 
+        // TODO: if this returns an error, zebrad panics elsewhere?
```

Comment:
If this returns an error, Zebra drops the modified chain, and returns an error for the block:

Chain handling:

During a chain update or fork:
https://github.com/ZcashFoundation/zebra/blob/main/zebra-state/src/service/non_finalized_state.rs#L143

When adding a new chain:
https://github.com/ZcashFoundation/zebra/blob/main/zebra-state/src/service/non_finalized_state.rs#L181

The error should propagate right back to the full block verifier, and then to the syncer or inbound downloader, which drops the block. If there's a panic, let's fix that code?

Reply:
```suggestion
```